hedera-node/src/main/java/com/hedera/services/state/virtual/UniqueTokenKey.java

```diff
@@ -0,0 +1,267 @@
+     * - 8) indicating number of bytes needed for the entity number. - The lower 4-bits will contain
+     * a number (0 - 8) indicating number of bytes needed for the serial number. - (variable, 0 to 8
+     * bytes) the non-leading zero-bytes representing the entity number. - (variable, 0 to 8 bytes)
+     * the non-leading zero-bytes representing the serial number.
+     */
+    public static final int ESTIMATED_SIZE_BYTES = Long.BYTES + Long.BYTES + 1;
+
+    /**
+     * Constructs a UniqueTokenKey from an NftId instance.
+     *
+     * @param nftId the NftId to create a UniqueTokenKey from.
+     * @return a new instance of a UniqueTokenKey corresponding to NftId.
+     */
+    public static UniqueTokenKey from(final NftId nftId) {
+        return new UniqueTokenKey(nftId.num(), nftId.serialNo());
+    }
+
+    public static UniqueTokenKey from(final NftNumPair nftNumPair) {
+        return new UniqueTokenKey(nftNumPair.tokenNum(), nftNumPair.serialNum());
+    }
+
+    /** The entity number of the token. */
+    private long entityNum;
+
+    /** Serial number of the token. */
+    private long tokenSerial;
+
+    /** Hashcode will be updated whenever tokenNum changes. */
+    private int hashCode;
+
+    public UniqueTokenKey() {}
+
+    public UniqueTokenKey(final long entityNum, final long tokenSerial) {
+        setTokenId(entityNum, tokenSerial);
+    }
+
+    public long getNum() {
+        return entityNum;
+    }
+
+    public long getTokenSerial() {
+        return tokenSerial;
+    }
+
+    private void setTokenId(final long entityNum, final long tokenSerial) {
+        this.entityNum = entityNum;
+        this.tokenSerial = tokenSerial;
+        // Consider using NonCryptographicHashing.hash64(long1, long2) when made available.
+        this.hashCode = Objects.hash(entityNum, tokenSerial);
+    }
+
+    private static int computeNonZeroBytes(final long value) {
+        // The value returned from this will range in [0, 8].
+        if (value == 0) {
+            return 0;
+        }
+
+        // Max value here is (64 - 0)/8 = 8
+        // Min value here is ceil((64 - 63)/8) = 1
+        final var nonZeroBits = Long.SIZE - Long.numberOfLeadingZeros(value);
+        return (nonZeroBits / 8) + Math.min(1, nonZeroBits % 8);
+    }
+
+    @Override
+    public int compareTo(@NotNull final UniqueTokenKey other) {
+        if (this == other) {
+            return 0;
+        }
+        // Sort by entity num first, followed by token serial number.
+        if (this.entityNum == other.entityNum) {
+            return Long.compare(this.tokenSerial, other.tokenSerial);
+        }
+        return Long.compare(this.entityNum, other.entityNum);
+    }
+
+    /* package */ interface ByteConsumer {
+        void accept(byte b) throws IOException;
+    }
+
+    private static byte packLengths(final int upper, final int lower) {
+        return (byte) ((upper << 4) | (lower & 0x0F));
```

Comment:
do we need to worry about big/small endian ?

Reply:
No. Java is always big-endian.